server/src/main/java/org/elasticsearch/action/search/CCSSingleCoordinatorSearchProgressListener.java

```diff
@@ -0,0 +1,363 @@
+                    break; // safety check to make sure it hasn't hit a terminal FAILED/SKIPPED state where timeouts don't matter
+                }
+                SearchResponse.Cluster updated = new SearchResponse.Cluster(
+                    curr.getClusterAlias(),
+                    curr.getIndexExpression(),
+                    curr.isSkipUnavailable(),
+                    curr.getStatus(),
+                    curr.getTotalShards(),
+                    curr.getSuccessfulShards(),
+                    curr.getSkippedShards(),
+                    curr.getFailedShards(),
+                    curr.getFailures(),
+                    curr.getTook(),
+                    true  // set timedOut flag to true
+                );
+                swapped = clusterRef.compareAndSet(curr, updated);
+            } while (swapped == false);
+        }
+    }
+
+    /**
+     * Executed when a shard reports a query failure.
+     *
+     * @param shardIndex The index of the shard in the list provided by {@link SearchProgressListener#onListShards})}.
+     * @param shardTarget The last shard target that thrown an exception.
+     * @param e The cause of the failure.
+     */
+    @Override
+    public void onQueryFailure(int shardIndex, SearchShardTarget shardTarget, Exception e) {
+        if (clusters.hasClusterObjects() == false) {
+            return;
+        }
+        String clusterAlias = shardTarget.getClusterAlias();
+        if (clusterAlias == null) {
+            clusterAlias = RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY;
+        }
+        AtomicReference<SearchResponse.Cluster> clusterRef = clusters.getCluster(clusterAlias);
+        boolean swapped;
+        do {
+            TimeValue took = null;
+            SearchResponse.Cluster curr = clusterRef.get();
+            SearchResponse.Cluster.Status status = SearchResponse.Cluster.Status.RUNNING;
+            int numFailedShards = curr.getFailedShards() == null ? 1 : curr.getFailedShards() + 1;
+
+            assert curr.getTotalShards() != null : "total shards should be set on the Cluster but not for " + clusterAlias;
+            if (curr.getTotalShards() == numFailedShards) {
+                if (curr.isSkipUnavailable()) {
+                    status = SearchResponse.Cluster.Status.SKIPPED;
+                } else {
+                    status = SearchResponse.Cluster.Status.FAILED;
+                    // TODO in the fail-fast ticket, should we throw an exception here to stop the search?
+                }
+            } else if (curr.getTotalShards() == numFailedShards + curr.getSuccessfulShards()) {
+                status = SearchResponse.Cluster.Status.PARTIAL;
+                took = new TimeValue(timeProvider.buildTookInMillis());
+            }
+
+            List<ShardSearchFailure> failures = new ArrayList<>();
+            curr.getFailures().forEach(failures::add);
```

Comment:
Could use `CollectionUtils.appendToCopy` here to append to `failures` in a single line and not pass around mutable lists :)

Reply:
I could find a way to do this as a one line.

CollectionUtils doesn't seem to have a method like `.append(failures, curr.getFailures(), new ShardSearchFailure(e, shardTarget)`.

I did modify it to be this:
```java
            List<ShardSearchFailure> failures = new ArrayList<>();
            failures.addAll(curr.getFailures());
            failures.add(new ShardSearchFailure(e, shardTarget));
```
I could do:
```java
            List<ShardSearchFailure> failures = new ArrayList<>();
            failures.addAll(curr.getFailures());
            CollectionUtils.appendToCopy(failures, new ShardSearchFailure(e, shardTarget));
```
but that's the same number of lines of code and is less efficient since you are copying the `curr.getFailures()` list twice. Since this is a local variable I don't see that have a temporarily mutable list is problematic.

But let me know if there's a utility to do this as a one-liner that you know about.

Reply:
What's wrong with `CollectionUtils.appendToCopy(cure.getFailures(), new ShardSearchFailure(e, shardTarget));`?

Reply:
Oh, I see. Yes that method copies the incoming array and then returns an immutable `List`, which is actually what I want in the Cluster (an immutable List). Thanks. I'll change this code and a few other places in the codebase to do this.